### PR TITLE
HDDS-11617. Update hadoop to 3.4.1

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?
HDDS-11617. Update hadoop to 3.4.1

Please describe your PR in detail:
* Use my own docker image fork to test Hadoop 3.4.1.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11617

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by jojochuang** [comment]
There is still one more test failure because of protobuf version change. Working on that.

**Review by adoroszlai — COMMENTED** [review]
Thanks @jojochuang for continued efforts on this.

Suggested change to disable new S3A contract tests:

```diff
diff --git hadoop-ozone/dist/src/main/compose/common/s3a-test.sh hadoop-ozone/dist/src/main/compose/common/s3a-test.sh
index e803206846..72b0a31ce0 100644
--- hadoop-ozone/dist/src/main/compose/common/s3a-test.sh
+++ hadoop-ozone/dist/src/main/compose/common/s3a-test.sh
@@ -93,10 +93,13 @@ execute_s3a_tests() {
 EOF
 
   # Some tests are skipped due to known issues.
+  # - ITestS3AContractBulkDelete: HDDS-11661
+  # - ITestS3AContractCreate: HDDS-11663
   # - ITestS3AContractDistCp: HDDS-10616
+  # - ITestS3AContractMkdirWithCreatePerf: HDDS-11662
   # - ITestS3AContractRename: HDDS-10665
   mvn -B -V --fail-never --no-transfer-progress \
-    -Dtest='ITestS3AContract*, ITestS3ACommitterMRJob, !ITestS3AContractDistCp, !ITestS3AContractRename' \
+    -Dtest='ITestS3AContract*, ITestS3ACommitterMRJob, !ITestS3AContractBulkDelete, !ITestS3AContractCreate#testOverwrite*EmptyDirectory[*], !ITestS3AContractDistCp, !ITestS3AContractMkdirWithCreatePerf, !ITestS3AContractRename' \
     clean test
 
   local target="${RESULT_DIR}/junit/${bucket}/target"
```

File: hadoop-ozone/ozonefs-shaded/pom.xml [hunk 1]
```diff
@@ -78,6 +78,12 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protobuf-java</artifactId>
+      <version>2.5.0</version>
+      <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (jojochuang) [reply hadoop-ozone/ozonefs-shaded/pom.xml:85]: agreed.

**Comment by jojochuang** [comment]
Thanks for the review! Will update accordingly.

File: hadoop-ozone/ozonefs-shaded/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jojochuang) [reply hadoop-ozone/ozonefs-shaded/pom.xml:85]: we should also use this opportunity to drop protobuf 2.5 from Ozone. Will. do that in a separate jira.

File: hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/MetricsAsserts.java [hunk 2]
```diff
@@ -411,7 +412,7 @@ public static void assertQuantileGauges(String prefix,
   public static void assertQuantileGauges(String prefix,
       MetricsRecordBuilder rb, String valueName) {
     verify(rb).addGauge(eqName(info(prefix + "NumOps", "")), geq(0L));
-    for (Quantile q : MutableQuantiles.quantiles) {
+    for (Quantile q : QUANTILES.getQuantiles()) {
```
Comment (jojochuang) [inline hadoop-hdds/test-utils/src/main/java/org/apache/ozone/test/MetricsAsserts.java:415]: Workaround for https://issues.apache.org/jira/browse/HADOOP-19301

File: hadoop-ozone/dist/src/shell/conf/log4j.properties [hunk 3]
```diff
@@ -20,7 +20,7 @@ hadoop.log.dir=.
 hadoop.log.file=hadoop.log
 
 # Define the root logger to the system property "hadoop.root.logger".
-log4j.rootLogger=${hadoop.root.logger}, EventCounter
```
Comment (jojochuang) [inline hadoop-ozone/dist/src/shell/conf/log4j.properties:23]: https://issues.apache.org/jira/browse/HADOOP-17642

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/TestCSMMetrics.java [hunk 4]
```diff
@@ -154,6 +154,14 @@ static void runContainerStateMachineMetrics(
       assertCounter("NumContainerNotOpenVerifyFailures", 0L, metric);
       assertCounter("WriteChunkMsNumOps", 1L, metric);
 
+      applyTransactionLatency = getDoubleGauge(
```
Comment (jojochuang) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/TestCSMMetrics.java:157]: https://issues.apache.org/jira/browse/HADOOP-18502

File: hadoop-ozone/pom.xml [hunk 5]
```diff
@@ -386,7 +381,6 @@
         <module>ozonefs-shaded</module>
         <module>ozonefs-hadoop2</module>
         <module>ozonefs-hadoop3</module>
-        <module>ozonefs-hadoop3-client</module>
```
Comment (jojochuang) [inline hadoop-ozone/pom.xml:389]: revert HDDS-6926.

**Comment by adoroszlai** [comment]
With Hadoop 3.1.2 `hadoop dfs` commands from `rm` (Resource Manager) node are stuck in retry:

```
2024-11-08 09:55:48 INFO  RetryInvocationHandler:411 - java.lang.IllegalStateException, while invoking $Proxy11.submitRequest over nodeId=om2,nodeAddress=om2:9862 after 1 failover attempts. Trying to failover immediately.
```

at

```
"main" #1 prio=5 os_prio=0 tid=0x00007b1a08054000 nid=0x11d waiting on condition [0x00007b1a11db5000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.processWaitTimeAndRetryInfo(RetryInvocationHandler.java:130)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:107)
	- locked <0x00000006748192d0> (a org.apache.hadoop.io.retry.RetryInvocationHandler$Call)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:359)
	at com.sun.proxy.$Proxy11.submitRequest(Unknown Source)
	at org.apache.hadoop.ozone.om.protocolPB.Hadoop3OmTransport.submitRequest(Hadoop3OmTransport.java:80)
	at org.apache.hadoop.ozone.om.protocolPB.OzoneManagerProtocolClientSideTranslatorPB.submitRequest(OzoneManagerProtocolClientSideTranslatorPB.java:338)
	at org.apache.hadoop.ozone.om.protocolPB.OzoneManagerProtocolClientSideTranslatorPB.getServiceInfo(OzoneManagerProtocolClientSideTranslatorPB.java:1863)
	at org.apache.hadoop.ozone.client.rpc.RpcClient.<init>(RpcClient.java:273)
	at org.apache.hadoop.ozone.client.OzoneClientFactory.getClientProtocol(OzoneClientFactory.java:269)
	at org.apache.hadoop.ozone.client.OzoneClientFactory.getRpcClient(OzoneClientFactory.java:136)
	at org.apache.hadoop.fs.ozone.BasicOzoneClientAdapterImpl.<init>(BasicOzoneClientAdapterImpl.java:186)
	at org.apache.hadoop.fs.ozone.OzoneClientAdapterImpl.<init>(OzoneClientAdapterImpl.java:51)
	at org.apache.hadoop.fs.ozone.OzoneFileSystem.createAdapter(OzoneFileSystem.java:109)
	at org.apache.hadoop.fs.ozone.BasicOzoneFileSystem.initialize(BasicOzoneFileSystem.java:200)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3303)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:124)
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3352)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3320)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:479)
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:361)
	at org.apache.hadoop.fs.shell.PathData.expandAsGlob(PathData.java:325)
	at org.apache.hadoop.fs.shell.CommandWithDestination.getRemoteDestination(CommandWithDestination.java:195)
	at org.apache.hadoop.fs.shell.CopyCommands$Put.processOptions(CopyCommands.java:259)
	at org.apache.hadoop.fs.shell.Command.run(Command.java:175)
	at org.apache.hadoop.fs.FsShell.run(FsShell.java:328)
	at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:76)
	at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:90)
	at org.apache.hadoop.fs.FsShell.main(FsShell.java:391)
```


With Hadoop 3.4.1 it's failing with:

```
Exception in thread "main" java.lang.NoClassDefFoundError: com/google/protobuf/ServiceException
	at org.apache.hadoop.ipc.ProtobufRpcEngine.getProxy(ProtobufRpcEngine.java:114)
	at org.apache.hadoop.ipc.RPC.getProtocolProxy(RPC.java:674)
	at org.apache.hadoop.ipc.RPC.getProtocolProxy(RPC.java:639)
	at org.apache.hadoop.ozone.om.ha.OMFailoverProxyProviderBase.createOMProxy(OMFailoverProxyProviderBase.java:143)
	at org.apache.hadoop.ozone.om.ha.HadoopRpcOMFailoverProxyProvider.createOMProxy(HadoopRpcOMFailoverProxyProvider.java:145)
	at org.apache.hadoop.ozone.om.ha.HadoopRpcOMFailoverProxyProvider.getProxy(HadoopRpcOMFailoverProxyProvider.java:132)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$ProxyDescriptor.<init>(RetryInvocationHandler.java:202)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.<init>(RetryInvocationHandler.java:335)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.<init>(RetryInvocationHandler.java:329)
	at org.apache.hadoop.io.retry.RetryProxy.create(RetryProxy.java:61)
	at org.apache.hadoop.ozone.om.protocolPB.Hadoop3OmTransport.createRetryProxy(Hadoop3OmTransport.java:116)
	at org.apache.hadoop.ozone.om.protocolPB.Hadoop3OmTransport.<init>(Hadoop3OmTransport.java:73)
	at org.apache.hadoop.ozone.om.protocolPB.Hadoop3OmTransportFactory.createOmTransport(Hadoop3OmTransportFactory.java:33)
	at org.apache.hadoop.ozone.om.protocolPB.OmTransportFactory.create(OmTransportFactory.java:45)
	at org.apache.hadoop.ozone.client.rpc.RpcClient.createOmTransport(RpcClient.java:414)
	at org.apache.hadoop.ozone.client.rpc.RpcClient.<init>(RpcClient.java:261)
        ...
	at org.apache.hadoop.fs.FsShell.main(FsShell.java:390)
```

It looks like shading protobuf does not work, because we still rely on Hadoop jars outside of our fat FS jar (and classes in those jars do not know about our shaded protobuf).

_acceptance (MR)_ [passed](https://github.com/jojochuang/ozone/actions/runs/11735618372/job/32693998999) in your fork without actually executing any Hadoop / MapReduce tests.

**Comment by adoroszlai** [comment]
```
hadoop-3.1.2-hadoopfs-o3fs :: Test ozone fs with hadoopfs             | PASS |
...
hadoop-3.4.1-hadoopfs-ofs :: Test ozone fs with hadoopfs              | PASS |
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @jojochuang for the patch, LGTM.  Let's ask for a few more reviews.  We should also try the fat jar with Spark, it's not covered by any tests.

**Comment by adoroszlai** [comment]
> It looks like shading protobuf does not work, because we still rely on Hadoop jars outside of our fat FS jar (and classes in those jars do not know about our shaded protobuf).

By the same logic we need to keep `ozone-filesystem-hadoop3-client` (tested with Spark 3.4.4 locally).

**Comment by jojochuang** [comment]
another approach is to add protobuf 2.5 jar into the hadoop classpath here: https://github.com/apache/ozone/blob/master/hadoop-ozone/dist/src/main/compose/common/hadoop3.conf 

what do you think?

**Comment by adoroszlai** [comment]
> another approach is to add protobuf 2.5 jar into the hadoop classpath here: 
> https://github.com/apache/ozone/blob/6dd566f08c67d8bc26c323c44af207e4f010c8f9/hadoop-ozone/dist/src/main/compose/common/hadoop3.conf#L19

`ozone-filesystem-*` is supposed to be a single drop-in jar, so I think it's better to include protobuf, too.

**Comment by jojochuang** [comment]
LGTM. But let's wait in case anyone else wants to chime in.

**Review by adoroszlai — APPROVED** [review]

**Review by slfan1989 — APPROVED** [review]

**Comment by jojochuang** [comment]
Merged. Thanks @adoroszlai for the help! and @slfan1989 for reviews.

**Pull request merged**